### PR TITLE
Torified webview in react-native

### LINKED ISSUE #42: Android/iOS app. Torified webview in react-native.
It might be a low hanging fruit to create a tiny react-native webview app wrapping the current webapp. It might also be possible to use react-native-tor https://github.com/Sifir-io/react-native-tor, so the user does not need TOR Browser nor Orbot, but only the .apk.

This could be a good start towards something more versatile and smartphone friendly.

### PR DESCRIPTION
**Pull request by KoalaSat**
Fixes #42 

We previously refactored the front-end to have a high level API client we can easily switch https://github.com/Reckless-Satoshi/robosats/pull/242 and created a single component for avatars so static image files can be easily managed https://github.com/Reckless-Satoshi/robosats/pull/251

This PR includes:
- New client on the front-end that communicates with react-native.
- Tor client on react-native that listen for requests, fetch and returns them back to webview.
- Other fixes found related with running the front-end on Android, and closely related to HTTP requests.

image_0 (1)
Screenshot taken from an Android device, all the data is being obtained through the Tor daemon.

## Issues found on Android

Details of this list: https://github.com/Reckless-Satoshi/robosats/issues/42#issuecomment-1250668802

- [x] Fetching internal files
Including extra static files on the app will also help to make sure the translation files are on the same version as the JS on the app. My intial approach was to run a static server, but it's not possible because of https://github.com/futurepress/react-native-static-server/issues/115, so I just included `i18n` on the mobile webpack bundle
**Solution**: https://github.com/Reckless-Satoshi/robosats/pull/247/files#diff-a7def86cc36a2178b4ee07b000c53d49e04068db4bc85967c0758578a2f37700R44 
- [x] External sources
Using the `request` function I manage to do XHR requests and obtain the base64 of external files. 
**Solution:** https://github.com/Reckless-Satoshi/robosats/pull/247/files#diff-ab78113fffebdaf893575b9bdd85086680e9553cf9f9f10bec0a1dc0e11e1c35R55
- [x] MUI Avatar not loading
After https://github.com/Reckless-Satoshi/robosats/pull/251 it was way easier to asynchronously manage the image `src` load, I also implemented an in-memory cache to avoid loading the image every time the `src` is requested (only for the Android build).
**Solution:** https://github.com/Reckless-Satoshi/robosats/pull/247/files#diff-35aa11cfb6d388c4d5c62abbc4b3bc0281ea8d19bc1ec800f0a0f47a59fedbfbR47
- [x] react-native-tor does not implement PUT
https://github.com/Sifir-io/react-native-tor/issues/51 We decided to not implement it.
- [x] react-native-tor crashes
Following this workaround seems to fix the issue https://github.com/Sifir-io/react-native-tor/issues/30#issuecomment-891060913 . At least for my case, it also increased considerably the general performance on Tor requests.
- [ ] Cookies not available on local files
https://github.com/react-native-webview/react-native-webview/issues/2643 I started working on it, but I ended up modifying +15 files on the web front-end, so I think it's better to just disable cookies on Android and fix it later. **HEADS UP!** Because of this, accessing the order page will fail, yet the requests are still being done
```
Error: Request Response Code (400): {"bad_request":"You must have a robot avatar to see the order details"}
```

## Changes to test on Web front-end
- Routes
- User Gen Main Avatar
- Translations

## Remaining work for a fully functional Android app
- Cookies
- Communicate with clipboard
- Websockets

## Further Considerations
- I'm not sure at all about the security risks on [this command](https://github.com/Reckless-Satoshi/robosats/pull/247/files#diff-64dbdc1262d05a2fb7d6fd141eab466d004060268ec220f2c39616c72f240d04R12) but I couldn't find a way to make the JS injection work without it. 
- There is a `JavaBinder` error that repeats over and over and I couldn't find a solution for it, looks like something related to the build size
image

### CODE REVIEW

File: frontend/src/components/UserGenPage.js [hunk 1]
```diff
@@ -231,20 +229,13 @@ class UserGenPage extends Component {
...
-                    />
-                  </div>
-                </Tooltip>
+                <RobotAvatar
+                  nickname={this.state.nickname}
+                  style={{ maxWidth: 200 * fontSizeFactor, maxHeight: 200 * fontSizeFactor }}
+                  smooth={true}
+                  avatarClass=""
+                  tooltip={t('This is your trading avatar')}
+                />
```
Comment (KoalaSat) [inline frontend/src/components/UserGenPage.js:238]: I missed this one on the Refactor, should be checked on frontend

File: frontend/src/components/UnsafeAlert.js [hunk 2]
```diff
@@ -14,17 +14,21 @@ class UnsafeAlert extends Component {
...
-    http.open('HEAD', `${location.protocol}//${this.getHost()}/selfhosted`, false);
-    http.send();
-    return http.status == 200;
+    try {
+      http.open('HEAD', `${location.protocol}//${this.getHost()}/selfhosted`, false);
+      http.send();
+      return http.status === 200;
+    } catch {
+      return false;
+    }
```
Comment (KoalaSat) [inline frontend/src/components/UnsafeAlert.js:31]: A gentle `try...catch` did the job :)

**Comment by Reckless-Satoshi** [comment]
Other than the main avatar looking different (style missing) everything else seems to work as intended in web. Ready to merge to create a pre-release with a .apk for testing. 

Submit a LN invoice from a proxy wallet for 2M Sats. Thanks a lot for figuring out one of the biggest hurdles so far during the development of RoboSats! :rocket:

File: mobile/services/Tor/index.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+class TorClient {
+  baseUrl: string;
+  daemon: ReturnType<typeof Tor>;
+
+  constructor() {
+    this.baseUrl = 'http://robosats6tkf3eva7x2voqso3a5wcorsnw34jveyxfqi2fu7oyheasid.onion';
+    this.daemon = Tor({
+      stopDaemonOnBackground: false,
+      numberConcurrentRequests: 0,
```
Comment (KoalaSat) [inline mobile/services/Tor/index.ts:11]: @Reckless-Satoshi According to https://github.com/Sifir-io/react-native-tor/blob/ed9dc6f1c474c3aa1ae7512f7437343159a46480/src/index.tsx#L349 we are safe to set this to 0 (we are using v8) and it should resolve the error you got. I tested loading all pages and didn't get any error.

**Comment by KoalaSat** [comment]
> Other than the main avatar looking different (style missing) everything else seems to work as intended in web. Ready to merge to create a pre-release with a .apk for testing.
> 
> Submit a LN invoice from a proxy wallet for 2M Sats. Thanks a lot for figuring out one of the biggest hurdles so far during the development of RoboSats! 🚀

Has been fun!

**Comment by KoalaSat** [comment]
```
lnbc20m1p3nx4rhpp5v8kasn3myn5k6sg8rfawtel8pj64948ch0mqr3wdx3r4zj6f738qdq5g9kxy7fqd9h8vmmfvdjscqzpgxqyz5vqsp5cl5f0skrxn3dmeuhn6f7q9zsaxtha3398xeh5qaqhl83q69samxs9qyyssqe4nvqxpljaaprx9uy9j8ke6yl96l9dj8w8ekf27zvht7hq2ecqkh7uyqc4rzyt67yz0utgsmnqcvwqxjx2ljfhguc3wtqluy5glu03sp2qec2w
```

**Comment by Reckless-Satoshi** [comment]
> ```
> lnbc20m1p3nx4rhpp5v8kasn3myn5k6sg8rfawtel8pj64948ch0mqr3wdx3r4zj6f738qdq5g9kxy7fqd9h8vmmfvdjscqzpgxqyz5vqsp5cl5f0skrxn3dmeuhn6f7q9zsaxtha3398xeh5qaqhl83q69samxs9qyyssqe4nvqxpljaaprx9uy9j8ke6yl96l9dj8w8ekf27zvht7hq2ecqkh7uyqc4rzyt67yz0utgsmnqcvwqxjx2ljfhguc3wtqluy5glu03sp2qec2w
> ```

3837663333336464643462343531353332346162383631326635353661313764

**Pull request merged**